src/Avalonia.Controls/Primitives/Popup.cs

```diff
@@ -277,6 +277,11 @@ public void Open()
         /// </summary>
         public void Close()
         {
+            if (_topLevel is null)
```

Comment:
Couple of things I wonder about exiting early here:

- Will not unsubscribing from `_popupHost.TemplateApplied` cause a problem? At first I din't think it should as it's created and used solely by the `Popup` instance. But then I wondered what would happen if the popup is re-attached to the tree and re-opened. Would that cause a double subscription?
- Similarly `IsOpen` is going to remain `true`. Will that cause a problem if the popup is reattached?

Reply:
I think code in the `Popup` is generally quite a mess. Looking at code in `Open` I can tell that if `_topLevel` is `null` then `_popupHost` must be `null` as well.

About the second point - you are right it might happen if `Open` somehow fails and then somebody calls `Close`. I could perhaps set it to `false` even in the early exit branch.

Reply:
Yeah I think this class is due a refactor, if you fancy getting your hands dirty...